File: DSA/Stack/Code/InfixToPostfix.java
// Infix to Postfix in Java

public class InfixToPostfix {
    static int precedence(char c) {
        if (c == '+' || c == '-') return 1;
        if (c == '*' || c == '/') return 2;
        if (c == '^') return 3;
        return 0;
    }
    
    static boolean isOperator(char c) {
        return (c == '+' || c == '-' || c == '*' || c == '/' || c == '^');
    }
    
    static String infixToPostfix(String infix) {
        StringBuilder postfix = new StringBuilder();
        java.util.Stack<Character> stack = new java.util.Stack<>();
        
        for (char c : infix.toCharArray()) {
            if (Character.isLetterOrDigit(c)) {
                postfix.append(c);
            } else if (c == '(') {
                stack.push(c);
            } else if (c == ')') {
                while (!stack.isEmpty() && stack.peek() != '(') {
                    postfix.append(stack.pop());
                }
                stack.pop(); // Remove '('
            } else if (isOperator(c)) {
                while (!stack.isEmpty() && precedence(stack.peek()) >= precedence(c)) {
                    postfix.append(stack.pop());
                }
                stack.push(c);
            }
        }
        
        while (!stack.isEmpty()) {
            postfix.append(stack.pop());
        }
        
        return postfix.toString();
    }
    
    public static void main(String[] args) {
        String infix = "a+b*c-d/e";
        
        System.out.println("=== Infix to Postfix ===\n");
        System.out.println("Infix Expression: " + infix);
        
        String postfix = infixToPostfix(infix);
        
        System.out.println("Postfix Expression: " + postfix + "\n");
        
        System.out.println("Algorithm: Shunting-yard");
        System.out.println("Complexity: O(n)");
    }
}

File: DSA/Stack/Code/InfixToPostfix.cs
// Infix to Postfix in C#

using System;
using System.Collections.Generic;
using System.Text;

class InfixToPostfix {
    static int Precedence(char c) {
        if (c == '+' || c == '-') return 1;
        if (c == '*' || c == '/') return 2;
        if (c == '^') return 3;
        return 0;
    }
    
    static bool IsOperator(char c) {
        return (c == '+' || c == '-' || c == '*' || c == '/' || c == '^');
    }
    
    static string InfixToPostfixConversion(string infix) {
        StringBuilder postfix = new StringBuilder();
        Stack<char> stack = new Stack<char>();
        
        foreach (char c in infix) {
            if (char.IsLetterOrDigit(c)) {
                postfix.Append(c);
            } else if (c == '(') {
                stack.Push(c);
            } else if (c == ')') {
                while (stack.Count > 0 && stack.Peek() != '(') {
                    postfix.Append(stack.Pop());
                }
                stack.Pop(); // Remove '('
            } else if (IsOperator(c)) {
                while (stack.Count > 0 && Precedence(stack.Peek()) >= Precedence(c)) {
                    postfix.Append(stack.Pop());
                }
                stack.Push(c);
            }
        }
        
        while (stack.Count > 0) {
            postfix.Append(stack.Pop());
        }
        
        return postfix.ToString();
    }
    
    static void Main() {
        string infix = "a+b*c-d/e";
        
        Console.WriteLine("=== Infix to Postfix ===\n");
        Console.WriteLine("Infix Expression: " + infix);
        
        string postfix = InfixToPostfixConversion(infix);
        
        Console.WriteLine("Postfix Expression: " + postfix + "\n");
        
        Console.WriteLine("Algorithm: Shunting-yard");
        Console.WriteLine("Complexity: O(n)");
    }
}
